cache/ringbuffer.go

```diff
@@ -14,102 +14,164 @@
 
 package cache
 
-import (
-	"fmt"
-
-	clientv3 "go.etcd.io/etcd/client/v3"
-)
-
-type ringBuffer struct {
-	buffer []batch
+type ringBuffer[T any] struct {
+	buffer []entry[T]
 	// head is the index immediately after the last non-empty entry in the buffer (i.e., the next write position).
-	head, tail, size int
+	head, tail, entryCount int
 }
 
-// batch groups all events that share one ModRevision.
-type batch struct {
-	rev    int64
-	events []*clientv3.Event
+type entry[T any] struct {
+	revision int64
+	items    []T
 }
 
 type KeyPredicate = func([]byte) bool
 
-func newRingBuffer(capacity int) *ringBuffer {
+func newRingBuffer[T any](capacity int) *ringBuffer[T] {
 	// assume capacity > 0 – validated by Cache
-	return &ringBuffer{
-		buffer: make([]batch, capacity),
+	return &ringBuffer[T]{
+		buffer: make([]entry[T], capacity),
 	}
 }
 
-func (r *ringBuffer) Append(events []*clientv3.Event) {
+type RevisionOf[T any] func(T) int64
+
+func (r *ringBuffer[T]) Append(items []T, revOf RevisionOf[T]) {
 	start := 0
-	for end := 1; end < len(events); end++ {
-		if events[end].Kv.ModRevision != events[start].Kv.ModRevision {
-			r.append(batch{
-				rev:    events[start].Kv.ModRevision,
-				events: events[start:end],
+	for end := 1; end < len(items); end++ {
+		if revOf(items[end]) != revOf(items[start]) {
+			r.append(entry[T]{
+				revision: revOf(items[start]),
+				items:    items[start:end],
 			})
 			start = end
 		}
 	}
-	if start < len(events) {
-		r.append(batch{
-			rev:    events[start].Kv.ModRevision,
-			events: events[start:],
+	if start < len(items) {
+		r.append(entry[T]{
+			revision: revOf(items[start]),
+			items:    items[start:],
 		})
 	}
 }
 
-func (r *ringBuffer) append(b batch) {
-	if len(b.events) == 0 {
+func (r *ringBuffer[T]) append(entry entry[T]) {
+	if len(entry.items) == 0 {
 		return
 	}
-	if r.size == len(r.buffer) {
+	if r.entryCount == len(r.buffer) {
 		r.tail = (r.tail + 1) % len(r.buffer)
 	} else {
-		r.size++
+		r.entryCount++
 	}
-	r.buffer[r.head] = b
+	r.buffer[r.head] = entry
 	r.head = (r.head + 1) % len(r.buffer)
 }
 
-// Filter returns all events in the buffer whose ModRevision is >= minRev.
-// TODO: use binary search on the ring buffer to locate the first entry >= nextRev instead of a full scan
-func (r *ringBuffer) Filter(minRev int64) (eventBatches [][]*clientv3.Event) {
-	eventBatches = make([][]*clientv3.Event, 0, r.size)
+// GetFromRev returns all entries with revision >= minRev
+func (r *ringBuffer[T]) GetFromRevision(minRev int64) [][]T {
```

Comment:
Maybe we could use naming from btree implementation, for this function we could use name `AscendGreatherOrEqual`

https://github.com/google/btree/blob/aeba20f7a1e1315badec4eca4fdc9f754f5f880a/btree_generic.go#L766